### PR TITLE
Handle user:group specification in Dockerfile

### PR DESCRIPTION
**Pull request by csrwng**
When checking whether a user is in the allowed range, a USER directive with uid:group is now allowed.

### CODE REVIEW

**Comment by csrwng** [comment]
@bparees ptal

File: pkg/docker/util.go [hunk 1]
```diff
@@ -214,13 +215,40 @@ func CheckAllowedUser(d Docker, imageName string, uids user.RangeList, isOnbuild
...
+// IsOnbuildAllowed checks a list of Docker ONBUILD instructions for
+// user directives. It ensures that any users specified by the directives
+// falls within the specified range list of users.
+func isOnbuildAllowed(directives []string, allowed *user.RangeList) bool {
+	for _, line := range directives {
+		parts := dockerLineDelim.Split(line, 2)
+		if strings.ToLower(parts[0]) != "user" {
+			continue
+		}
+		uname := extractUser(parts[1])
```
Comment (mfojtik) [inline pkg/docker/util.go]: is this also trimming the uname?

Comment (csrwng) [inline pkg/docker/util.go]: it's not... it should. Will fix

Comment (csrwng) [inline pkg/docker/util.go]: fixed

**Comment by bparees** [comment]
lgtm.

File: pkg/docker/util_test.go [hunk 2]
```diff
@@ -78,6 +78,32 @@ func TestCheckAllowedUser(t *testing.T) {
 			onbuild:     []string{"USER 501", "VOLUME /data"},
 			expectErr:   false,
 		},
+		{
+			name:        "AllowedUIDs is set, numeric user with group",
+			allowedUIDs: rangeList("1-"),
+			user:        "5:5000",
+			expectErr:   false,
```
Comment (GrahamDumpleton) [inline pkg/docker/util_test.go]: Should add a test for:

```
5:users
```

That is, named group if want to ensure coverage of all cases.

**Comment by mfojtik** [comment]
LGTM, pending fixing @GrahamDumpleton check

**Comment by GrahamDumpleton** [comment]
Hold off merging this. As per comments on original issue, this may be pointless as deployer may not use the group anyway. If that is the case, need to work if deployer should honour the group in Docker image data.

**Comment by csrwng** [comment]
@GrahamDumpleton I believe we still need the fix for s2i not to reject images unnecessarily. We're simply not interpreting the user spec in a Dockerfile correctly right now.

File: pkg/docker/util_test.go (hunk 2 above)
Comment (csrwng) [inline pkg/docker/util_test.go]: added, thx

**Comment by bparees** [comment]
@csrwng +1, i think this fix is as good as it can get from an s2i perspective.  lgtm.

**Comment by GrahamDumpleton** [comment]
I have created https://github.com/openshift/origin/issues/8618 for wider issue of changing Kubernetes/OpenShift to honour the group. This change in this patch is still useful in context of standalone S2I usage, but images produced still aren't going to necessarily run properly on OpenShift.

**Comment by bparees** [comment]
@csrwng @GrahamDumpleton are we in agreement that this PR is as complete as it's going to get?

**Comment by csrwng** [comment]
yes

**Comment by bparees** [comment]
[merge]

**Pull request merged**